src/Form/SettingsForm.php

```diff
@@ -0,0 +1,271 @@
+      '#default_value' => $defaultValues[self::ORGANISATION_SERVICE_ENDPOINT_REFERENCE] ?? NULL,
+      '#description' => $this->t('Probably "http://stoettesystemerne.dk/service/organisation/3", but it may very well change in the future.'),
+    ];
+
+    $form[self::ORGANISATION_TEST_LEDER_ROLLE_UUID] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Leder rolle uuid test'),
+      '#required' => TRUE,
+      '#default_value' => $defaultValues[self::ORGANISATION_TEST_LEDER_ROLLE_UUID] ?? NULL,
+    ];
+
+    $form[self::ORGANISATION_PROD_LEDER_ROLLE_UUID] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Leder rolle uuid produktion'),
+      '#required' => TRUE,
+      '#default_value' => $defaultValues[self::ORGANISATION_PROD_LEDER_ROLLE_UUID] ?? NULL,
+    ];
+
+    $form['actions']['#type'] = 'actions';
+
+    $form['actions']['submit'] = [
+      '#type' => 'submit',
+      '#value' => $this->t('Save settings'),
+    ];
+
+    $form['actions']['testCertificate'] = [
+      '#type' => 'submit',
+      '#name' => 'testCertificate',
+      '#value' => $this->t('Test certificate'),
+    ];
+
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   *
+   * @phpstan-param array<string, mixed> $form
+   */
+  public function validateForm(array &$form, FormStateInterface $formState): void {
+    $triggeringElement = $formState->getTriggeringElement();
+    if ('testCertificate' === ($triggeringElement['#name'] ?? NULL)) {
+      return;
+    }
+
+    $values = $formState->getValues();
+
+    // Validate cache expiration.
+    try {
+      $cacheExpirationOptions = explode(',', $values[self::CACHE_EXPIRATION]);
+      foreach ($cacheExpirationOptions as $cacheExpirationOption) {
+        new \DateTime($cacheExpirationOption);
+      }
+    }
+    catch (\Exception $exception) {
+      $formState->setErrorByName(self::CACHE_EXPIRATION, $this->t('Invalid cache expiration: %cache_expiration', ['%cache_expiration' => $values[self::CACHE_EXPIRATION]]));
+    }
+
+    if (CertificateLocatorHelper::LOCATOR_TYPE_FILE_SYSTEM === $values['certificate']['locator_type']) {
+      $path = $values['certificate'][CertificateLocatorHelper::LOCATOR_TYPE_FILE_SYSTEM]['path'] ?? NULL;
+      if (!file_exists($path)) {
+        $formState->setErrorByName('certificate][file_system][path', $this->t('Invalid certificate path: %path', ['%path' => $path]));
+      }
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   *
+   * @phpstan-param array<string, mixed> $form
+   */
+  public function submitForm(array &$form, FormStateInterface $formState): void {
+    $triggeringElement = $formState->getTriggeringElement();
+    if ('testCertificate' === ($triggeringElement['#name'] ?? NULL)) {
+      $this->testCertificate();
+      return;
+    }
+
+    $values = $formState->getValues();
+    foreach ($this->settings->getKeys() as $key) {
+      if (array_key_exists($key, $values)) {
+        $this->settings->set($key, $values[$key]);
+      }
+    }
```

Comment:
I know where this is coming from, but it's bad way to do this. Be more explicit like in https://github.com/itk-dev/os2forms_digital_post/blob/feature/DW-277-SF1601/src/Form/SettingsForm.php#L233-L254.

Reply:
Fixed.